wcfsetup/install/files/lib/data/I18nDatabaseObjectList.class.php

```diff
@@ -0,0 +1,72 @@
+<?php
+namespace wcf\data;
+use wcf\system\exception\SystemException;
+use wcf\system\WCF;
+use wcf\util\StringUtil;
+
+/**
+ * Abstract class for a list of database objects with better sorting of i18n-based columns.
+ *
+ * @author	Florian Gail
+ * @copyright	2001-2017 WoltLab GmbH
+ * @license	GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package	WoltLabSuite\Core\Data
+ * @since	3.1
+ */
+abstract class I18nDatabaseObjectList extends DatabaseObjectList {
+	/**
+	 * Array of column names, that are eventually filled with language items.
+	 * Those additional joins are going to slow down the system; you should use this
+	 * class only when it's really needed.
+	 * The key represents the original field name.
+	 * The value represents the new field name containing the localized field-value.
+	 *
+	 * @example   [ 'title' => 'titleSortField' ]
+	 *
+	 * @var string[]
+	 */
+	public $i18nFields = [];
+	
+	/**
+	 * ID of the language that should be used
+	 * the value defaults to the current user's language
+	 *
+	 * @var integer
+	 */
+	public $languageID;
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function __construct() {
+		parent::__construct();
+		
+		if ($this->languageID === null) {
+			$this->languageID = WCF::getUser()->languageID;
+		}
+		
+		if (!empty($this->i18nFields)) {
+			if (count($this->i18nFields) != count(array_flip($this->i18nFields))) {
+				throw new SystemException("'".$this->className."::\$i18nFields' should unique values.");
+			}
+			
+			foreach ($this->i18nFields as $key => $value) {
+				if (is_numeric($key)) {
+					throw new SystemException("'" . $this->className . "::\$i18nFields' should contain alphanumeric key-value pairs. '" . $value . "' has an empty or numeric key.");
+				}
+				if (!preg_match('/^([0-9a-zA-Z]+)$/', $value)) {
```

Comment:
I suggest a single if with this regex: `/^[a-z][a-zA-Z0-9]*$/` and the following message:

> Array keys of MyDBO::$i18nFields must start with a small letter and consist of letters and number only.